Windows build

mtsokol opened the pull request:
Here's a draft PR with Windows build support.

Some comments:
- I think it's worth considering to switch to `setup-bazel` and `setup-python` Actions for Bazel and Python setup for all environments, rather than downloading and installing them in the script.
  I couldn't make it work with the current pyenv and bazel installation - switching to GitHub Actions worked out of the box.
- The Bazel setup complained about `Module.bazel` dependencies versions - I upgraded them.
- It seems that the wheel build is so far blocked by `array-record` - package: https://pypi.org/project/array-record/#history. The current setup requires `0.7.2` but Windows wheels for it is missing and the last one goes back to `0.4.1`: https://github.com/mtsokol/grain/actions/runs/15418522140/job/43387085894.
  I think that first we need to make sure that all dependencies have Windows distributions.

iindyk commented:
thanks Mateusz!

> I think it's worth considering to switch to setup-bazel and setup-python Actions for Bazel and Python setup for all environments, rather than downloading and installing them in the script.

Agree, would you be able to try this out? maybe in a separate PR to make things easier

> The Bazel setup complained about Module.bazel dependencies versions - I upgraded them.

this part is tricky, especially the protobuf dependency version. Downstream users are very sensitive to it. In particular, TensorFlow has pretty old protobuf dependency and using the newer ones may result into a crash in the env with grain + tensorflow. We wouldn't want that ideally. Would it be possible to not upgrade the dependencies yet? were the complaints errors?

> It seems that as of now the wheel build is blocked by array-record - package

yeah, there's an ongoing work to fix array-record build (it's broken at main), once that is in, or folks will add Windows build. For now, we shouldn't block on that (i.e. let the install fail, at least verify that we can build wheels).

mtsokol commented:
> Agree, would you be able to try this out? maybe in a separate PR to make things easier

Sure! I'll take care of it. 

> Would it be possible to not upgrade the dependencies yet? were the complaints errors?

I think I upgraded only those that otherwise gave a dependency resolve error. But I can rerun it and paste details here - maybe there's a workaround for it.

mtsokol commented:
@iindyk some updates:

- Opened refactor PR: https://github.com/google/grain/pull/900 <- let's merge it first, then I'll rebase this one.
- Indeed dependencies doesn't need to be updated. Once I figured out how to pin Bazel version on installation `USE_BAZEL_VERSION: "7.1.1"` the current setup is fine, so I reverted these changes.
- About the `array-record`: Currently `bazel build` fails with [missing array-record for Windows](https://github.com/mtsokol/grain/actions/runs/15559058746/job/43806617036). `BUILD` file requires it, so letting it fail means, `bazel build` fails, so we can't build the wheel anyway (i.e. we rsync `bazel-bin/grain` outputs to the wheel build directory). Am I missing anything?

mtsokol commented on grain/_src/python/data_sources.py:
```diff
@@ -26,19 +26,19 @@
 
 from absl import logging
-import array_record.python.array_record_data_source as array_record
 from etils import epath
```
I couldn't use `from etils import epy; epy.lazy_imports()` as `Union` didn't allow `epy`'s returned type.

mtsokol commented on grain/oss/build_whl.sh:
```diff
@@ -29,10 +29,20 @@ main() {
+    *)
+      INCLUDE_EXT="*.so"
+      # Also reduce noise during build.
+      write_to_bazelrc "build --cxxopt=-Wno-deprecated-declarations --host_cxxopt=-Wno-deprecated-declarations"
```
I searched it and AFAIK There are no counterparts in MSVC.

mtsokol commented on grain/BUILD:
```diff
@@ -30,7 +30,10 @@ py_library(
         "transforms.py",
     ],
-    data = ["//grain/_src/python/experimental/index_shuffle/python:index_shuffle_module.so"],
+    data = select({
+        "@platforms//os:windows": ["//grain/_src/python/experimental/index_shuffle/python:index_shuffle_module.pyd"],
+        "//conditions:default": ["//grain/_src/python/experimental/index_shuffle/python:index_shuffle_module.so"],
+    }),
```
Does Bazel allow to do it in a more concise manner? (Move path to a variable an glue the extension separately?)

iindyk replied:
not sure, I think this is ok

iindyk commented on grain/_src/python/dataset/transformations/shuffle.py:
```diff
@@ -18,9 +18,12 @@
-from grain._src.python.experimental.index_shuffle.python import index_shuffle_module as index_shuffle
+
+with epy.lazy_imports():
+  from grain._src.python.experimental.index_shuffle.python import index_shuffle_module as index_shuffle
```
does this import not work on Windows?

iindyk commented on grain/oss/build_whl.sh:
```diff
@@ -71,10 +81,13 @@ main() {
-  rsync -avm -L --exclude="__pycache__/*" grain "${TMPDIR}"
-  rsync -avm -L  --include="*.so" --include="*_pb2.py" \
+
+  rsync -avm -L --exclude="__pycache__/*" grain tmp_folder
```
could you pls add a comment about the intermediate folder here?

mtsokol commented:
@iindyk I addressed your comments. In the follow-up I will add skips for 10 tests that fail on windows (e.g. due to unix-like paths in assert strings or `compile_pip_requirements` tests that check for `array_record`) and move JAX tests to pytest, so that CI with "Run unit tests" option will pass.

iindyk commented:
thanks! merged in https://github.com/google/grain/commit/9772d55228a0b32ed72fdee893c12a31030b056f
